### PR TITLE
IGNITE-14222 Exclude evicted partitions from re-encryption.

### PR DESCRIPTION
**Pull request by xtern**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java [hunk 1]
```diff
@@ -442,74 +485,88 @@ public long remainingPagesCount() {
...
+                    int pagesCnt = Math.min(batchSize, cnt - off);
 
-                limiter.acquire(pagesCnt);
+                    limiter.acquire(pagesCnt);
 
-                synchronized (this) {
-                    if (isDone() || !parts.contains(partId))
-                        break;
+                    synchronized (this) {
+                        if (isDone() || evicted(partId))
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java:536]: Let's move evicted check under checkpoint read lock. There can be a race again in the current implementation with the same outcome as fixed by this patch.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/Checkpointer.java [hunk 2]
```diff
@@ -713,31 +713,34 @@ public void schedulePartitionDestroy(@Nullable CacheGroupContext grpCtx, int grp
...
+                canceled = req.cancel();
         }
 
-        if (req != null)
-            req.waitCompleted();
+        if (req != null) {
+            if (!canceled)
+                req.waitCompleted();
+            else if (log.isInfoEnabled())
+                log.info("Partition file destroy has cancelled [grpId=" + grpId + ", partId=" + partId + "]");
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/Checkpointer.java]: Done, thanks

File: modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java (hunk 1 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/Checkpointer.java [hunk 3]
```diff
@@ -713,31 +713,34 @@ public void schedulePartitionDestroy(@Nullable CacheGroupContext grpCtx, int grp
...
+            req = scheduledCp.getDestroyQueue().removeRequest(grpId, partId);
 
-        CheckpointProgressImpl cur;
+            if (req == null) {
+                CheckpointProgressImpl cur = curCpProgress;
 
-        synchronized (this) {
-            cur = curCpProgress;
+                if (cur != null)
+                    req = cur.getDestroyQueue().removeRequest(grpId, partId);
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/Checkpointer.java]: Added deletion from both scheduled and current cp progress.

**Pull request closed without merging**